# deng-lab/viroprofiler: Citations

| Software      | Module                              | License     |
| ------------- | ----------------------------------- | ----------- |
| metaSPAdes    | Assembly                            | NA          |
| vRhyme        | Binning                             | GPL v3      |
| Phamb         | Binning                             | MIT License |
| CheckV        | Virus detection and QC              | BSD license |
| VirSorter2    | Virus detection                     | GPL v2      |
| DeepVirFinder | Virus detection                     | USC-RL v1.0 |
| VIBRANT       | Virus detection and gene annotation | GPL v3      |
| DRAM          | Gene annotation                     | GPL v3      |
| eggnog-mapper | Gene annotation                     | GPL v3      |
| abricate      | Gene annotation                     | GPL v2      |
| MMseqs2       | Taxonomy assignment                 | GPL v3      |
| vConTACT2     | Taxonomy assignment                 | GPL v3      |
| Bacphlip      | Lifestyle prediction                | MIT License |
| Replidec      | Lifestyle prediction                | MIT License |
| iPHoP         | Host prediction                     | GPL v3      |
| CoverM        | Abundance estimation                | GPL v3      |
| Kraken2       | Taxonomy assignment                 | MIT License |
| Bracken       | Abundance estimation                | GPL v3      |

## [nf-core](https://pubmed.ncbi.nlm.nih.gov/32055031/)

> Ewels, Philip A., et al. "The nf-core framework for community-curated bioinformatics pipelines." Nature biotechnology 38.3 (2020): 276-278.

## [Nextflow](https://pubmed.ncbi.nlm.nih.gov/28398311/)

> Di Tommaso, Paolo, et al. "Nextflow enables reproducible computational workflows." Nature biotechnology 35.4 (2017): 316-319.

## Pipeline tools

- [FastQC](https://www.bioinformatics.babraham.ac.uk/projects/fastqc/)

- [fastp](https://)
  > Chen, Shifu, et al. "fastp: an ultra-fast all-in-one FASTQ preprocessor." Bioinformatics 34.17 (2018): i884-i890.

- [BBMap](https://jgi.doe.gov/data-and-tools/software-tools/bbtools/bb-tools-user-guide/bbmap-guide/)
  > Bushnell, Brian. BBMap: a fast, accurate, splice-aware aligner. No. LBNL-7065E. Lawrence Berkeley National Lab.(LBNL), Berkeley, CA (United States), 2014.

- [metaSPAdes](https://github.com/ablab/spades)
  > Nurk, Sergey, et al. "metaSPAdes: a new versatile metagenomic assembler." Genome research 27.5 (2017): 824-834.

- [Bowtie 2](https://github.com/BenLangmead/bowtie2)
  > Langmead, Ben, and Steven L. Salzberg. "Fast gapped-read alignment with Bowtie 2." Nature methods 9.4 (2012): 357-359.

- [CoverM](https://github.com/wwood/CoverM)

- [CheckV](https://bitbucket.org/berkeleylab/CheckV)
  > Nayfach, Stephen, et al. "CheckV assesses the quality and completeness of metagenome-assembled viral genomes." Nature biotechnology 39.5 (2021): 578-585.

- [VirSorter2](https://github.com/jiarong/VirSorter2)
  > Guo, Jiarong, et al. "VirSorter2: a multi-classifier, expert-guided approach to detect diverse DNA and RNA viruses." Microbiome 9.1 (2021): 1-13.

- [DeepVirFinder](https://github.com/jessieren/DeepVirFinder)
  > Ren, Jie, et al. "Identifying viruses from metagenomic data using deep learning." Quantitative Biology 8.1 (2020): 64-77.

- [VIBRANT](https://github.com/AnantharamanLab/VIBRANT)
  > Kieft, Kristopher, Zhichao Zhou, and Karthik Anantharaman. "VIBRANT: automated recovery, annotation and curation of microbial viruses, and evaluation of viral community function from genomic sequences." Microbiome 8.1 (2020): 1-23.

- [vRhyme](https://github.com/AnantharamanLab/vRhyme)
  > Kieft K, Adams A, Salamzade R, et al. vRhyme enables binning of viral genomes from metagenomes[J]. Nucleic Acids Research, 2022, 50(14): e83-e83.

- [Phamb](https://github.com/RasmussenLab/phamb)
  > Johansen, Joachim, et al. "Genome binning of viral entities from bulk metagenomics data." Nature communications 13.1 (2022): 1-12.

- [DRAM-v](https://github.com/WrightonLabCSU/DRAM)
  > Shaffer, Michael, et al. "DRAM for distilling microbial metabolism to automate the curation of microbiome function." Nucleic acids research 48.16 (2020): 8883-8900.

- [vConTACT2](https://bitbucket.org/MAVERICLab/vcontact2)
  > Bin Jang, Ho, et al. "Taxonomic assignment of uncultivated prokaryotic virus genomes is enabled by gene-sharing networks." Nature biotechnology 37.6 (2019): 632-639.

- [MMseqs2](https://github.com/soedinglab/MMseqs2)
  > Steinegger, Martin, and Johannes Söding. "MMseqs2 enables sensitive protein sequence searching for the analysis of massive data sets." Nature biotechnology 35.11 (2017): 1026-1028.
  > Mirdita, Milot, et al. "Fast and sensitive taxonomic assignment to metagenomic contigs." Bioinformatics 37.18 (2021): 3029-3031.

- [eggNOG-mapper](https://github.com/eggnogdb/eggnog-mapper)
  > Cantalapiedra C P, Hernández-Plaza A, Letunic I, et al. eggNOG-mapper v2: functional annotation, orthology assignments, and domain prediction at the metagenomic scale[J]. Molecular biology and evolution, 2021, 38(12): 5825-5829.

- [abricate](https://github.com/tseemann/abricate)

- [iPHoP](https://bitbucket.org/srouxjgi/iphop/src/main/)
  > Roux S, Camargo AP, Coutinho FH, Dabdoub SM, Dutilh BE, Nayfach S, Tritt A. iPHoP: An integrated machine learning framework to maximize host prediction for metagenome-derived viruses of archaea and bacteria. PLoS biology. 2023 Apr 21;21(4):e3002083.

- [bacphlip](https://github.com/adamhockenberry/bacphlip)
  > Hockenberry A J, Wilke C O. BACPHLIP: predicting bacteriophage lifestyle from conserved protein domains[J]. PeerJ, 2021, 9: e11396.

- [Replidec](https://github.com/deng-lab/Replidec)
  > Peng, Xue, et al. "Replidec-Use naive Bayes classifier to identify virus lifecycle from metagenomics data." bioRxiv (2022): 2022-07.

- [Kraken2](https://github.com/DerrickWood/kraken2)
  > Wood, Derrick E., Jennifer Lu, and Ben Langmead. "Improved metagenomic analysis with Kraken 2." Genome biology 20 (2019): 1-13.

- [Bracken](https://github.com/jenniferlu717/Bracken)
  > Lu, Jennifer, et al. "Bracken: estimating species abundance in metagenomics data." PeerJ Computer Science 3 (2017): e104.

- [MultiQC](https://pubmed.ncbi.nlm.nih.gov/27312411/)
  > Ewels, Philip, et al. "MultiQC: summarize analysis results for multiple tools and samples in a single report." Bioinformatics 32.19 (2016): 3047-3048.

## Software packaging/containerisation tools

- [Anaconda](https://anaconda.com)
  > Anaconda Software Distribution. Computer software. Vers. 2-2.4.0. Anaconda, Nov. 2016. Web.

- [Bioconda](https://pubmed.ncbi.nlm.nih.gov/29967506/)
  > Grüning B, Dale R, Sjödin A, et al. Bioconda: sustainable and comprehensive software distribution for the life sciences[J]. Nature methods, 2018, 15(7): 475-476.

- [BioContainers](https://pubmed.ncbi.nlm.nih.gov/28379341/)
  > da Veiga Leprevost F, Grüning B A, Alves Aflitos S, et al. BioContainers: an open-source and community-driven framework for software standardization[J]. Bioinformatics, 2017, 33(16): 2580-2582.

- [Docker](https://dl.acm.org/doi/10.5555/2600239.2600241)
  > Merkel D. Docker: lightweight linux containers for consistent development and deployment[J]. Linux j, 2014, 239(2): 2.

- [Singularity](https://pubmed.ncbi.nlm.nih.gov/28494014/)
  > Kurtzer G M, Sochat V, Bauer M W. Singularity: Scientific containers for mobility of compute[J]. PloS one, 2017, 12(5): e0177459.

